### PR TITLE
Monitors

### PR DESCRIPTION
**Pull request by SimonHeybrock**
I think I would like to add monitor handling to the examples we will be preparing for the dev. meeting. Not sure how yet, either we can get monitors from Mantid if available and also convert them, or we just add some fake ones, for demo purposes.

Anyway, here is some prep work for this: Add Python example of adding monitors, and add an attribute tag for monitors as well. This also highlighted some overload resolution issue in Pybind11, which is also fixed here.

### CODE REVIEW

File: python/test/test_dataset.py [hunk 1]
```diff
@@ -482,6 +492,38 @@ def test_Workspace2D_example(self):
...
+        profile[Coord.Y] = ([Dim.Y], [-0.02, -0.01, 0.0, 0.01, 0.02])
+        profile[Data.Value] = ([Dim.Y, Dim.X], (4,4))
+        # Monitors can also be attributes, so they are not required to match in operations
+        d[Attr.Monitor, "beam-profile"] = ([], profile)
+
+        # Add histogram-mode transmission monitor
+        transmission = Dataset()
+        transmission[Coord.Energy] = ([Dim.Energy], np.arange(9))
+        transmission[Data.Value] = ([Dim.Energy], np.random.exponential(size=8))
+        d[Coord.Monitor, "transmission"] = ([], ())
```
Comment (OwenArnold) [inline python/test/test_dataset.py]: These seem to be more usage examples than "unit tests" and would eventually need to be re-homed accordingly. Also looks like this is 3 tests in one, maybe different named methods?

Comment (OwenArnold) [reply python/test/test_dataset.py]: Would be nice to see these monitors used/accessed as well as defined.

File: python/test/test_variable.py [hunk 2]
```diff
@@ -48,10 +48,13 @@ def test_create_dtype(self):
         var = Variable(Coord.X, [Dim.X], (4,), dtype=np.dtype(np.int32))
         self.assertEqual(var.numpy.dtype, np.dtype(np.int32))
 
-    def test_coord_set_name_fails(self):
+    def test_coord_set_name(self):
         var = Variable(Coord.X, [Dim.X], np.arange(4))
-        with self.assertRaises(RuntimeError):
-            var.name = "data"
+        # We can set a name for coordinates, but not that operations will typically
```
Comment (OwenArnold) [inline python/test/test_variable.py]: What has been the driver for named coordinates?

File: python/test/test_dataset.py [hunk 3]
```diff
@@ -482,6 +492,38 @@ def test_Workspace2D_example(self):
...
+
+        # Add data with uncertainties
+        d[Data.Value, "sample1"] = ([Dim.Spectrum, Dim.Tof], np.random.exponential(size=100*8).reshape([100, 8]))
+        d[Data.Variance, "sample1"] = d[Data.Value, "sample1"]
+
+        # Add event-mode beam-status monitor
+        status = Dataset()
+        status[Data.Tof] = ([Dim.Event], np.random.exponential(size=1000))
+        status[Data.PulseTime] = ([Dim.Event], np.random.exponential(size=1000))
+        d[Coord.Monitor, "beam-status"] = ([], status)
```
Comment (OwenArnold) [inline python/test/test_dataset.py]: What about a histogram monitor. Still treated as a nested dataset, i.e. make use of different binning?

File: python/test/test_dataset.py (hunk 1 above)
Comment (SimonHeybrock) [reply python/test/test_dataset.py]: I am currently working on an actual larger example where monitors are really used.

File: python/test/test_variable.py (hunk 2 above)
Comment (SimonHeybrock) [reply python/test/test_variable.py]: - Avoids either one level of nesting for monitors or gets rid of an extra auxiliary dimensions `Dim::Monitors`, which is confusing. Provides named access without requiring a `DatasetIndex` (compare the old example https://github.com/mantidproject/dataset/blob/master/doc/design.md#examples-cpp).
- Coordinates that do not correspond to dimensions (those that are not "dimension coordinates") are more flexible. Consider storing data from a number of temperature sensors. Without names we would need `Coord::Temperature1`, `Coord::Temperature2`, ... which totally defeats the purpose. `(Coord::Temperature, "sample")` and `(Coord::Temperature, "cryostat")` are much clearer.
- Less code and special cases in the implementation?

File: python/test/test_dataset.py (hunk 3 above)
Comment (SimonHeybrock) [reply python/test/test_dataset.py]: Yes. It can be different if required.

**Pull request merged**